File: idealingua-v1/idealingua-v1-test-harness/src/main/csharp/FixtureSeeder.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using Newtonsoft.Json;

namespace IdealinguaV1Harness.CSharpDriver {
    /// <summary>
    /// One-shot seeder: constructs typed values and writes their canonical Newtonsoft-encoded
    /// JSON to wire-fixtures/csharp/.  Run via:
    ///   dotnet bin/Debug/net9.0/Driver.dll seed &lt;repoRoot&gt;
    /// Kept as a permanent diagnostic tool for re-baselining.
    /// </summary>
    public static class FixtureSeeder {
        public static void Seed(string repoRoot) {
            var fixturesRoot = Path.Combine(repoRoot, "idealingua-v1", "idealingua-v1-test-defs", "wire-fixtures", "csharp");
            Console.Error.WriteLine($"[seeder] writing fixtures to: {fixturesRoot}");

            var fixtures = BuildFixtures();
            foreach (var (wireId, scenario, value) in fixtures) {
                if (!Dispatch.Entries.TryGetValue(wireId, out var entry)) {
                    Console.Error.WriteLine($"[seeder] WARN: no dispatch entry for {wireId}");
                    continue;
                }
                var json = entry.Serialize(value);
                var dir  = Path.Combine(fixturesRoot, wireId);
                Directory.CreateDirectory(dir);
                var path = Path.Combine(dir, scenario + ".json");
                File.WriteAllText(path, json, new System.Text.UTF8Encoding(encoderShouldEmitUTF8Identifier: false));
                Console.Error.WriteLine($"[seeder] wrote {wireId}/{scenario}.json  →  {json}");
            }
            Console.Error.WriteLine("[seeder] done.");
        }

        private static List<(string wireId, string scenario, object value)> BuildFixtures() {
            var list = new List<(string, string, object)>();

            // ── Plain DTOs ──────────────────────────────────────────────────────────
            list.Add((
                "idltest.dtofields.Point", "basic",
                new Idltest.Dtofields.Point(10, 20, "abc", "point-1", 3, 4, "of", true)
            ));
            list.Add((
                "idltest.dtofields.Point", "extra",
                new Idltest.Dtofields.Point(0, 100, "xyz", "origin", 0, 0, "", false)
            ));

            list.Add((
                "idltest.dtofields.NullableObj", "basic",
                new Idltest.Dtofields.NullableObj(99)
            ));

            // Optional present: No = NullableObj(42)
            list.Add((
                "idltest.dtofields.OptionalObj", "with-some",
                new Idltest.Dtofields.OptionalObj(new Idltest.Dtofields.NullableObj(42))
            ));
            // Optional absent: No = null → serialises as {}
            list.Add((
                "idltest.dtofields.OptionalObj", "with-none",
                new Idltest.Dtofields.OptionalObj(null)
            ));

            // ListObj: list of two NullableObjs
            list.Add((
                "idltest.dtofields.ListObj", "basic",
                new Idltest.Dtofields.ListObj(new List<Idltest.Dtofields.NullableObj> {
                    new Idltest.Dtofields.NullableObj(1),
                    new Idltest.Dtofields.NullableObj(2),
                })
            ));

            // ── Identifiers ─────────────────────────────────────────────────────────
            // BucketID: App=3a7f0c12..., Bucket="my-bucket", User=0a1b2c3d...
            var appGuid    = new Guid("3a7f0c12-1234-5678-9abc-fedcba987654");
            var userGuid   = new Guid("0a1b2c3d-4e5f-6071-8293-a4b5c6d7e8f9");
            var uid2       = new Guid("aaaabbbb-cccc-dddd-eeee-ffffaaaabbbb");
            var bucket1    = new Idltest.Identifiers.BucketID(appGuid, userGuid, "my-bucket");
            var bucket2    = new Idltest.Identifiers.BucketID(appGuid, userGuid, "test-bucket");
            var bucket3    = new Idltest.Identifiers.BucketID(appGuid, userGuid, "b1");

            list.Add((
                "idltest.identifiers.BucketID", "basic",
                bucket1
            ));

            // UserId: Company=0a1b2c3d..., Value=3a7f0c12...
            // ToString() → "UserId#" + company + ":" + value
            var userId = new Idltest.Identifiers.UserId(appGuid, userGuid);
            list.Add((
                "idltest.identifiers.UserId", "basic",
                userId
            ));

            // DepartmentEnum
            list.Add((
                "idltest.identifiers.DepartmentEnum", "basic",
                Idltest.Identifiers.DepartmentEnum.Engineering
            ));

            // UserWithEnumId: Company=0a1b2c3d..., Dept=Sales, Value=3a7f0c12...
            // ToString() → "UserWithEnumId#" + company + ":" + dept + ":" + value
            var userWithEnumId = new Idltest.Identifiers.UserWithEnumId(appGuid, userGuid, Idltest.Identifiers.DepartmentEnum.Sales);
            list.Add((
                "idltest.identifiers.UserWithEnumId", "basic",
                userWithEnumId
            ));

            // ComplexID: Bucket=bucket2, User=userWithEnumIdE, I32=42, Uid=uid2, Str="hello"
            // where userWithEnumIdE uses Engineering dept (to match Scala fixture)
            var userWithEnumIdE = new Idltest.Identifiers.UserWithEnumId(appGuid, userGuid, Idltest.Identifiers.DepartmentEnum.Engineering);
            var complexId = new Idltest.Identifiers.ComplexID(bucket2, userWithEnumIdE, 42, uid2, "hello");
            list.Add((
                "idltest.identifiers.ComplexID", "basic",
                complexId
            ));

            // KVIDGeneric: map "key1" → bucket3
            var kvMap = new Dictionary<string, Idltest.Identifiers.BucketID> { { "key1", bucket3 } };
            list.Add((
                "idltest.identifiers.KVIDGeneric", "basic",
                new Idltest.Identifiers.KVIDGeneric(kvMap)
            ));

            // ── ADTs ────────────────────────────────────────────────────────────────
            list.Add((
                "idltest.algebraics.AdtTester", "as-ComplexAdt",
                (object)new Idltest.Algebraics.AdtTester.ComplexAdt(
                    new Idltest.Algebraics.ComplexAdt(new Idltest.Algebraics.AdtTestID("alpha"))
                )
            ));
            list.Add((
                "idltest.algebraics.AdtTester", "as-ComplexAdt2",
                (object)new Idltest.Algebraics.AdtTester.ComplexAdt2(
                    new Idltest.Algebraics.ComplexAdt2(new Idltest.Algebraics.AdtTestID("beta"))
                )
            ));

            // AdtWithInterface: AFace branch with AFaceStruct(a=7)
            list.Add((
                "idltest.algebraics.AdtWithInterface", "basic",
                (object)new Idltest.Algebraics.AdtWithInterface.AFace(new Idltest.Algebraics.AFaceStruct(7))
            ));

            // ── Interface impls ─────────────────────────────────────────────────────
            // WithCovariance.Struct: bare struct with field = CovariantStruct
            list.Add((
                "idltest.inheritance.WithCovariance.Struct", "basic",
                new Idltest.Inheritance.WithCovarianceStruct(
                    new Idltest.Inheritance.CovariantStruct()
                )
            ));

            // Empty.Struct: bare empty struct
            list.Add((
                "idltest.inheritance.Empty.Struct", "empty",
                new Idltest.Inheritance.EmptyStruct()
            ));

            // ── JSONLike ADT ─────────────────────────────────────────────────────────
            list.Add((
                "idltest.json.JSONLike", "basic",
                (object)new Idltest.Json.JSONLike.JLString(new Idltest.Json.JLString("hello"))
            ));

            // ── Cross-domain ─────────────────────────────────────────────────────────
            list.Add((
                "idltest.phase.Name_incoming", "basic",
                new Idltest.Phase.Name_incoming("test-name")
            ));

            // ── AllTypes.Struct (BC=N: ts → +00:00, unsigned native, float 1.5/−2.25) ──
            // ts:      UTC zoned  → TszDefault = "yyyy-MM-ddTHH:mm:ss.fffzzz".
            //          C# ParseExact with literal-Z format produces DateTimeKind.Unspecified,
            //          so re-serialize uses TszDefault → "+00:00" on UTC host (this host).
            //          We construct the ts value by parsing "+00:00" to get the stable round-trip form.
            // tsuni:   UTC univ   → TsuDefault = "yyyy-MM-ddTHH:mm:ss.fffZ"   → "Z"
            // tslocal: local      → TslDefault = "yyyy-MM-ddTHH:mm:ss.fff"    → no zone
            // uint*:   native positive values
            // Parse ts from the stable +00:00 form so seeder emits what the driver round-trips to.
            var tsUtc   = DateTime.ParseExact("2025-01-15T10:30:45.000+00:00",
                              IRT.Marshaller.JsonNetTimeFormats.Tsz,
                              System.Globalization.CultureInfo.InvariantCulture,
                              System.Globalization.DateTimeStyles.None);
            var tsLocal = new DateTime(2025, 1, 15, 10, 30, 45, DateTimeKind.Unspecified);
            var tsUni   = new DateTime(2025, 1, 15, 10, 30, 45, DateTimeKind.Utc);
            var time    = new TimeSpan(0, 10, 30, 45, 0);
            var date    = new DateTime(2025, 1, 15);
            var uuid    = new Guid("3a7f0c12-1234-5678-9abc-fedcba987654");
            var enumMap = new Dictionary<string, Izumi.Test.Domain01.GoAliasEnumTest> {
                { "k1", Izumi.Test.Domain01.GoAliasEnumTest.Val1 }
            };
            list.Add((
                "izumi.test.domain01.AllTypes.Struct", "basic",
                new Izumi.Test.Domain01.AllTypesStruct(
                    b: true,
                    s: "hello",
                    int8: 42,
                    int16: 1000,
                    int32: 100000,
                    int64: -9007199254740993L,
                    f: 1.5f,
                    d: -2.25,
                    uuid: uuid,
                    ts: tsUtc,
                    tslocal: tsLocal,
                    tsuni: tsUni,
                    time: time,
                    date: date,
                    uint8: 200,
                    uint16: 60000,
                    uint32: 4000000000u,
                    uint64: 9007199254740993ul,
                    list: new List<Izumi.Test.Domain01.AllTypes>(),
                    another: new List<Izumi.Test.Domain01.AllTypes>(),
                    selfMap: new Dictionary<string, Izumi.Test.Domain01.AllTypes>(),
                    enumMap: enumMap,
                    option: null,
                    selfSet: new List<Izumi.Test.Domain01.AllTypes>(),
                    optionDate: null,
                    optionTime: null
                )
            ));

            return list;
        }
    }
}

File: idealingua-v1/idealingua-v1-test-harness/src/main/typescript/FixtureSeeder.ts
// PR-03.3a FixtureSeeder.ts — generates TS wire fixtures from actual encoder output.
// Usage: npx tsx FixtureSeeder.ts <repoRoot>
// Mirrors PR-03.2's FixtureSeeder.scala: kept as a permanent diagnostic tool.
import * as fs from 'fs';
import * as path from 'path';

import { Point } from 'idltest/dtofields/Point';
import { NullableObj } from 'idltest/dtofields/NullableObj';
import { OptionalObj } from 'idltest/dtofields/OptionalObj';
import { ListObj } from 'idltest/dtofields/ListObj';
import { ComplexID } from 'idltest/identifiers/ComplexID';
import { UserId } from 'idltest/identifiers/UserId';
import { BucketID } from 'idltest/identifiers/BucketID';
import { KVIDGeneric } from 'idltest/identifiers/KVIDGeneric';
import { DepartmentEnum } from 'idltest/identifiers/DepartmentEnum';
import { UserWithEnumId } from 'idltest/identifiers/UserWithEnumId';
import { ComplexAdt } from 'idltest/algebraics/ComplexAdt';
import { ComplexAdt2 } from 'idltest/algebraics/ComplexAdt2';
import { AdtTesterHelpers } from 'idltest/algebraics/AdtTester';
import { AdtWithInterfaceHelpers } from 'idltest/algebraics/AdtWithInterface';
import { AFaceStruct } from 'idltest/algebraics/AFace';
import { WithCovarianceStruct } from 'idltest/inheritance/WithCovariance';
import { CovariantStruct } from 'idltest/inheritance/Covariant';
import { EmptyStruct } from 'idltest/inheritance/Empty';
import { JLString } from 'idltest/json/JLString';
import { JSONLikeHelpers } from 'idltest/json/JSONLike';
// AllTypesStruct import skipped — formatter.ts moment() issue under tsx/esbuild; see B4 note below.
import { Name_incoming } from 'idltest/phase/Name_incoming';

const repoRoot = process.argv[2];
if (!repoRoot) {
  console.error('Usage: npx tsx FixtureSeeder.ts <repoRoot>');
  process.exit(1);
}

const fixturesRoot = path.join(repoRoot, 'idealingua-v1/idealingua-v1-test-defs/wire-fixtures/typescript');

function write(wireId: string, scenario: string, value: any): void {
  const dir = path.join(fixturesRoot, wireId);
  fs.mkdirSync(dir, { recursive: true });
  const outPath = path.join(dir, `${scenario}.json`);
  const json = JSON.stringify(value);
  fs.writeFileSync(outPath, json, { encoding: 'utf8' });
  console.log(`  wrote ${wireId}/${scenario}.json  (${json.length} bytes)`);
}

function serializeSingle(val: { serialize(): any }): any {
  return val.serialize();
}

console.log('FixtureSeeder: starting');

// ── B1: Point, NullableObj, Empty ─────────────────────────────────────────────

// 1a. Point basic
{
  const p = new Point({
    w: 10, h: 20, id: 'abc', name: 'point-1', x: 3, y: 4, ownfield: 'of', export: true
  });
  write('idltest.dtofields.Point', 'basic', p.serialize());
}

// 1b. Point extra
{
  const p = new Point({
    w: 0, h: 100, id: 'xyz', name: 'origin', x: 0, y: 0, ownfield: '', export: false
  });
  write('idltest.dtofields.Point', 'extra', p.serialize());
}

// 14. NullableObj
{
  const n = new NullableObj({ a: 99 });
  write('idltest.dtofields.NullableObj', 'basic', n.serialize());
}

// 15. Empty.Struct
{
  const e = new EmptyStruct({});
  write('idltest.inheritance.Empty.Struct', 'empty', e.serialize());
}

// ── B2: ADTs, WithCovariance, JSONLike ─────────────────────────────────────────

// 5a. AdtTester as-ComplexAdt
{
  const adt = new ComplexAdt({ id: 'AdtTestID#alpha' });
  write('idltest.algebraics.AdtTester', 'as-ComplexAdt', AdtTesterHelpers.serialize(adt));
}

// 5b. AdtTester as-ComplexAdt2
{
  const adt = new ComplexAdt2({ id: 'AdtTestID#beta' });
  write('idltest.algebraics.AdtTester', 'as-ComplexAdt2', AdtTesterHelpers.serialize(adt));
}

// 6. AdtWithInterface basic — AFace branch
{
  const aface = AFaceStruct.create({ 'idltest.algebraics.AFace.Struct': { a: 7 } });
  write('idltest.algebraics.AdtWithInterface', 'basic', AdtWithInterfaceHelpers.serialize(aface));
}

// 7. WithCovariance.Struct basic
{
  const cov = CovariantStruct.create({ 'idltest.inheritance.Covariant.Struct': {} });
  const wc = new WithCovarianceStruct({ field: { 'idltest.inheritance.Covariant.Struct': {} } });
  write('idltest.inheritance.WithCovariance.Struct', 'basic', wc.serialize());
}

// 21. JSONLike basic — JLString branch
{
  const jl = new JLString({ value: 'hello' });
  write('idltest.json.JSONLike', 'basic', JSONLikeHelpers.serialize(jl));
}

// ── B3: Optional, List, Map, Enum, phase ─────────────────────────────────────

// 8. OptionalObj with-some
{
  const o = new OptionalObj({ no: { a: 42 } });
  write('idltest.dtofields.OptionalObj', 'with-some', o.serialize());
}

// 9. OptionalObj with-none — TS drops the key (undefined not emitted by JSON.stringify)
{
  const o = new OptionalObj({} as any);
  write('idltest.dtofields.OptionalObj', 'with-none', o.serialize());
}

// 10. ListObj basic
{
  const l = new ListObj({ all: [{ a: 1 }, { a: 2 }] });
  write('idltest.dtofields.ListObj', 'basic', l.serialize());
}

// 11. KVIDGeneric basic
{
  const kv = new KVIDGeneric({
    test: { key1: 'BucketID#3a7f0c12-1234-5678-9abc-fedcba987654:b1:0a1b2c3d-4e5f-6071-8293-a4b5c6d7e8f9' }
  });
  write('idltest.identifiers.KVIDGeneric', 'basic', kv.serialize());
}

// 12. DepartmentEnum basic
{
  const e = DepartmentEnum.Engineering;
  write('idltest.identifiers.DepartmentEnum', 'basic', DepartmentEnum[e]);
}

// 13. UserWithEnumId basic
{
  const u = new UserWithEnumId('UserWithEnumId#0a1b2c3d-4e5f-6071-8293-a4b5c6d7e8f9:Sales:3a7f0c12-1234-5678-9abc-fedcba987654');
  write('idltest.identifiers.UserWithEnumId', 'basic', u.serialize());
}

// 29. Name_incoming basic
{
  const n = new Name_incoming({ name: 'test-name' });
  write('idltest.phase.Name_incoming', 'basic', n.serialize());
}

// ── B1 identifiers ───────────────────────────────────────────────────────────

// 2. ComplexID basic
{
  const c = new ComplexID('ComplexID#BucketID%233a7f0c12-1234-5678-9abc-fedcba987654%3Atest-bucket%3A0a1b2c3d-4e5f-6071-8293-a4b5c6d7e8f9:42:hello:aaaabbbb-cccc-dddd-eeee-ffffaaaabbbb:UserWithEnumId%230a1b2c3d-4e5f-6071-8293-a4b5c6d7e8f9%3AEngineering%3A3a7f0c12-1234-5678-9abc-fedcba987654');
  write('idltest.identifiers.ComplexID', 'basic', c.serialize());
}

// 3. UserId basic
{
  const u = new UserId('UserId#0a1b2c3d-4e5f-6071-8293-a4b5c6d7e8f9:3a7f0c12-1234-5678-9abc-fedcba987654');
  write('idltest.identifiers.UserId', 'basic', u.serialize());
}

// 3b. BucketID basic
{
  const b = new BucketID('BucketID#3a7f0c12-1234-5678-9abc-fedcba987654:my-bucket:0a1b2c3d-4e5f-6071-8293-a4b5c6d7e8f9');
  write('idltest.identifiers.BucketID', 'basic', b.serialize());
}

// ── B4: AllTypes ─────────────────────────────────────────────────────────────
// AllTypes.Struct is SKIPPED in TS fixtures (PR-03.3a).
//
// Root cause: formatter.ts uses `import * as moment from 'moment'`. Under tsx/esbuild CJS
// transform, `import * as` produces `var moment = __toESM(require("moment"))`, where
// __toESM wraps the CJS function in a namespace Object.create(Function.prototype), which
// is non-callable. Node 24 + tsx 4.21 do not set esModuleInterop default behaviour that
// would avoid this. Cannot fix without modifying formatter.ts (a golden file).
// Tracked as F-AllTypes-TS-skip; fix in PR-03.4 (daemon mode allows per-file tsconfig override)
// or when goldens are regenerated with esModuleInterop-aware formatter.

console.log('FixtureSeeder: done');
